allennlp/commands/train.py

```diff
@@ -152,6 +166,52 @@ def datasets_from_params(params: Params) -> Dict[str, Iterable[Instance]]:
 
     return datasets
 
+def create_serialization_dir(params: Params, serialization_dir: str) -> None:
+    """
+    This function creates the serialization directory if it doesn't exist.
+    If it already exists, then it verifies that we're continuing from a training with an identical configuration.
+    Parameters
+    ----------
+    params: Params, required.
+        A parameter object specifying an AllenNLP Experiment.
+    serialization_dir: str, required
+        The directory in which to save results and logs.
+    """
+    if os.path.exists(serialization_dir):
+        logger.info(f"Continuing from prior training at {serialization_dir}.")
+
+        continued_config_file = os.path.join(serialization_dir, CONFIG_NAME)
+        if not os.path.exists(continued_config_file):
+            logger.warning("The serialization directory already exists but doesn't contain a config.json.")
+        else:
+            loaded_params = Params.from_file(continued_config_file)
+
+            # Check whether any of the training configuration differs from the configuration we are resuming.
+            # If so, warn the user that training may fail.
+            fail = False
+            flat_params = params.as_flat_dict()
+            flat_loaded = loaded_params.as_flat_dict()
+            for key in flat_params.keys() - flat_loaded.keys():
+                logger.error(f"Key '{key}' found in training configuration but not in the serialization "
+                             f"directory we're continuing from.")
+                fail = True
+            for key in flat_loaded.keys() - flat_params.keys():
+                logger.error(f"Key '{key}' found in the serialization directory we're continuing from "
+                             f"but not in the training config.")
+                fail = True
+            for key in flat_params.keys():
```

Comment:
Making the config.json be identical seems like a very harsh requirement - for instance, you could imagine wanting to continue training with the same model on different data, or with a different optimiser/learning rate, all of which would require changing the configuration but are conceptually not errors. Perhaps we could check that the `model` section of the configuration was identical instead? Even then, I can imagine situations where you would want to alter something.

Reply:
@matt-gardner was your intention that we just check the model section in https://github.com/allenai/allennlp/issues/770?

Reply:
I think those use cases are better served with a different command, something like `fine-tune`.  In that case, you'd take a model archive and a new dataset, and continue training the model.  Trying to do too much here would make the `--continue` flag too hard to understand.

I understand the problem, though - calling this "continue" makes you think it's just a general mechanism for continuing training from some point.  Maybe `--recover` would be more appropriate for what's going on.

Reply:
Or we could have a `--force`.

Reply:
That'd be even more complicated...  My vote is for `--recover`, and leaving any other kind of "continue" semantics to another command.  Makes it really clear what the intended use case is here, which makes it easier to understand and maintain.